tensorflow/contrib/rnn/kernels/lstm_ops.h

```diff
@@ -0,0 +1,272 @@
+#include "tensorflow/core/framework/tensor_types.h"
+#include "tensorflow/core/platform/types.h"
+#include "tensorflow/stream_executor/stream.h"
+
+namespace tensorflow {
+
+#if GOOGLE_CUDA
+
+namespace {
+template <typename T>
+perftools::gputools::DeviceMemory<T> AsDeviceMemory(const T* cuda_memory) {
+  perftools::gputools::DeviceMemoryBase wrapped(const_cast<T*>(cuda_memory));
+  perftools::gputools::DeviceMemory<T> typed(wrapped);
+  return typed;
+}
+}  // namespace
+
+#endif  // GOOGLE_CUDA
+
+namespace functor {
+
+template <typename Device, bool USE_CUBLAS>
+struct LSTMCellBlockFprop {
+  void operator()(
+      perftools::gputools::Stream* stream, const Device& d,
+      const int batch_size, const int input_size, const int cell_size, const float forget_bias,
+      typename TTypes<float>::ConstMatrix x,
+      typename TTypes<float>::Matrix xh,
+      typename TTypes<float>::ConstMatrix states_prev,
+      typename TTypes<float>::ConstMatrix w,
+      typename TTypes<float>::ConstVec b,
+      typename TTypes<float>::Matrix h,
+      typename TTypes<float>::Matrix states) {
+    // Pointer offsets.
+    Eigen::array<int, 2> i_offsets  = {0, cell_size * 0};
+    Eigen::array<int, 2> cs_offsets = {0, cell_size * 1};
+    Eigen::array<int, 2> f_offsets  = {0, cell_size * 2};
+    Eigen::array<int, 2> o_offsets  = {0, cell_size * 3};
+    Eigen::array<int, 2> ci_offsets = {0, cell_size * 4};
+    Eigen::array<int, 2> co_offsets = {0, cell_size * 5};
+    Eigen::array<int, 2> h_offsets  = {0, cell_size * 6};
+
+    Eigen::array<int, 2> cell_extents = {batch_size, cell_size};
+
+    // xh = [x, h_prev].
+    Eigen::array<int, 2> xh_x_offsets = {0, 0};
+    Eigen::array<int, 2> xh_x_extents = {batch_size, input_size};
+    Eigen::array<int, 2> xh_h_offsets = {0, input_size};
+    Eigen::array<int, 2> xh_h_extents = {batch_size, cell_size};
+
+    auto h_prev =
+        states_prev.slice(h_offsets, cell_extents);
+
+    xh.slice(xh_x_offsets, xh_x_extents).device(d) = x;
+    xh.slice(xh_h_offsets, xh_h_extents).device(d) = h_prev;
+
+    Eigen::array<int, 2> states_offsets = {0, 0};
+    Eigen::array<int, 2> states_extents = {batch_size, cell_size * 4};
+
+    // states = xh * w + b
+    if (USE_CUBLAS) {
+#if GOOGLE_CUDA
+      states.slice(states_offsets, states_extents).device(d) =
+          b.broadcast(Eigen::array<int, 2>({batch_size, 1}));
+
+      auto kNoTranspose =
+          perftools::gputools::blas::Transpose::kNoTranspose;
+
+      const uint64 m = batch_size;
+      const uint64 k = input_size + cell_size;
+      const uint64 n = cell_size * 4;
+
+      auto a_ptr = AsDeviceMemory(xh.data());
+      auto b_ptr = AsDeviceMemory(w.data());
+      auto c_ptr = AsDeviceMemory(states.data());
+
+      bool blas_launch_status = stream->ThenBlasGemm(
+          kNoTranspose, kNoTranspose, n, m, k, 1.0f, b_ptr, n, a_ptr, k, 1.0f,
+          &c_ptr, cell_size * 7).ok();
+      CHECK(blas_launch_status);
```

Comment:
same here -- check stream status and set the context / return, and then check the status in the caller